File: apps/android/app/src/main/java/com/kelpie/browser/browser/SessionStore.kt
package com.kelpie.browser.browser

import android.content.Context

/** Persists and restores the set of open tab URLs across app launches. */
object SessionStore {
    private const val PREFS_NAME = "kelpie_prefs"
    private const val KEY_URLS = "session_tab_urls"
    private const val KEY_ACTIVE = "session_active_index"

    fun save(
        context: Context,
        tabs: List<BrowserTab>,
        activeId: String?,
    ) {
        val valid = tabs.filter { it.currentUrl.isNotEmpty() && !it.isStartPage }
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        if (valid.isEmpty()) {
            prefs
                .edit()
                .remove(KEY_URLS)
                .remove(KEY_ACTIVE)
                .apply()
            return
        }
        val urls = valid.map { it.currentUrl }
        val activeIndex = valid.indexOfFirst { it.id == activeId }.coerceAtLeast(0)
        prefs
            .edit()
            .putString(KEY_URLS, urls.joinToString("\n"))
            .putInt(KEY_ACTIVE, activeIndex)
            .apply()
    }

    fun load(context: Context): Pair<List<String>, Int>? {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val raw = prefs.getString(KEY_URLS, null) ?: return null
        val urls = raw.split("\n").filter { it.isNotEmpty() }
        if (urls.isEmpty()) return null
        val active = prefs.getInt(KEY_ACTIVE, 0)
        return urls to active
    }

    fun clear(context: Context) {
        context
            .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .remove(KEY_URLS)
            .remove(KEY_ACTIVE)
            .apply()
    }
}

File: apps/macos/Kelpie/Browser/SessionStore.swift
import Foundation

/// Persists and restores open tabs across app launches.
///
/// Tabs now carry a display name and an optional storage partition, so the
/// session needs more than a URL list. The record array is written as JSON
/// under a new key; the legacy URL-array key is still read once so an existing
/// install does not lose its tabs on the upgrade, and is cleared afterwards.
enum SessionStore {
    /// One persisted tab.
    struct TabRecord: Codable {
        var url: String
        var name: String?
        var partition: String?
        var persistent = true
        var isActive = false
    }

    private static let tabsKey = "sessionTabs"
    private static let legacyURLsKey = "sessionTabURLs"
    private static let legacyActiveIndexKey = "sessionActiveIndex"

    @MainActor
    static func save(tabs: [Tab], activeID: UUID?) {
        // A non-persistent partition's storage does not survive the process, so
        // restoring its tab would reopen an authenticated-looking tab with an
        // empty store. Drop those at save time; `restore` guards again in case
        // an older blob is still on disk.
        let valid = tabs.filter { !$0.currentURL.isEmpty && !$0.isStartPage && $0.persistent }
        guard !valid.isEmpty else {
            clear()
            return
        }
        let records = valid.map { tab in
            TabRecord(
                url: tab.currentURL,
                name: tab.name,
                partition: tab.partition,
                persistent: tab.persistent,
                isActive: tab.id == activeID
            )
        }
        guard let data = try? JSONEncoder().encode(records) else {
            print("[SessionStore] failed to encode \(records.count) tab record(s) — session not saved")
            return
        }
        UserDefaults.standard.set(data, forKey: tabsKey)
        UserDefaults.standard.removeObject(forKey: legacyURLsKey)
        UserDefaults.standard.removeObject(forKey: legacyActiveIndexKey)
    }

    static func load() -> [TabRecord] {
        if let data = UserDefaults.standard.data(forKey: tabsKey) {
            guard let records = try? JSONDecoder().decode([TabRecord].self, from: data) else {
                print("[SessionStore] tab records failed to decode — starting from a blank session")
                clear()
                return []
            }
            return records
        }
        return loadLegacy()
    }

    /// Reads the pre-partition format: a plain URL array plus an active index.
    private static func loadLegacy() -> [TabRecord] {
        guard let urls = UserDefaults.standard.stringArray(forKey: legacyURLsKey), !urls.isEmpty else {
            return []
        }
        let activeIndex = UserDefaults.standard.integer(forKey: legacyActiveIndexKey)
        return urls.enumerated().map { index, url in
            TabRecord(url: url, isActive: index == activeIndex)
        }
    }

    static func clear() {
        UserDefaults.standard.removeObject(forKey: tabsKey)
        UserDefaults.standard.removeObject(forKey: legacyURLsKey)
        UserDefaults.standard.removeObject(forKey: legacyActiveIndexKey)
    }
}
